### PR TITLE
Ensure parameters are available for failed cases

### LINKED ISSUE #2504: TestNG 7.4 & 7.3 data provider data cannot be populated to onTestStart hook when BeforeMethod is failed
### TestNG Version
7.4 & 7.3

### Expected behavior
When beforeMethod is failed, and in `ITestListener.onTestStart(ITestResult result)`, I execute `result.getParameters()`, I should get the parameters from DataProvider, even the beforeMethod/beforeClass is failed

### Actual behavior
ITestListener.onTestStart(ITestResult result) cannot get full parameters from Data provider when beforeMethod/beforeClass is failed. As the result, the we cannot generate correct Allure Report,  since the parameters are always change. 

when beforeMethod is failed, the parameter I got from ITestResult for test is [],  when beforeMethod is passed, the parameter is proper data from data provider.  

**Other facts:**
beforeMethodInvoke and onTestSkipped works well in this case.

### Is the issue reproductible on runner?

- [ ] Shell
- [x] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [ ] NetBeans

### Test case sample

```
public class RobinTest2 {

    @DataProvider
    public Object[][] data() {
        return new Object[][]{
                new Object[]{Colours.GREEN},
                new Object[]{Colours.RED},
                new Object[]{Colours.YELLOW},
                new Object[]{Colours.BLUE},
                new Object[]{Colours.BLACK}
        };
    }


    @BeforeClass
    void myBeforeClass() throws Exception {
    }

    @BeforeMethod
    void myBeforeMethod() throws Exception {
        throw new Exception("my exception");
    }

    @Test(dataProvider = "data")
    void testSuccess(Colours colour) {
        Assert.assertTrue(true);
    }
}
```

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #2504

Fixes #2504  .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: src/test/java/test/dataprovider/issue2504/SampleTestCaseListener.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package test.dataprovider.issue2504;
+
+import org.testng.ITestListener;
+import org.testng.ITestResult;
+
+public class SampleTestCaseListener implements ITestListener {
+
+  private Object[] parameters;
```
Comment (juherr) [inline src/test/java/test/dataprovider/issue2504/SampleTestCaseListener.java:8]: Should be a `List<Object[]>`

File: src/test/java/test/dataprovider/issue2504/SampleTestCaseListener.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+  @Override
+  public void onTestStart(ITestResult result) {
+    this.parameters = result.getParameters();
```
Comment (juherr) [inline src/test/java/test/dataprovider/issue2504/SampleTestCaseListener.java:12]: Add parameters into the list

File: src/test/java/test/dataprovider/issue2504/IssueTest.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  @Test(description = "GITHUB-2504")
+  public void ensureParametersCopiedOnConfigFailures() {
+    XmlTest xmltest = createXmlTest("2504_suite", "2504_test");
+    xmltest.setVerbose(2);
+    xmltest.setXmlClasses(Collections.singletonList(new XmlClass(SampleTestCase.class)));
+    TestNG testNG = create(Collections.singletonList(xmltest.getSuite()));
+    SampleTestCaseListener listener = new SampleTestCaseListener();
+    testNG.addListener(listener);
+    testNG.run();
+    Assertions.assertThat(listener.getParameters()).hasSize(1);
```
Comment (juherr) [inline src/test/java/test/dataprovider/issue2504/IssueTest.java:22]: map the list and then assert it contains only `1, 2, 3, 4, 5`

File: src/test/java/test/dataprovider/issue2504/SampleTestCaseListener.java (hunk 1 above)
Comment (krmahadevan) [reply src/test/java/test/dataprovider/issue2504/SampleTestCaseListener.java]: Fixed

File: src/test/java/test/dataprovider/issue2504/IssueTest.java (hunk 3 above)
Comment (krmahadevan) [reply src/test/java/test/dataprovider/issue2504/IssueTest.java]: Fixed

**Comment by krmahadevan** [comment]
@juherr - Review comments fixed. Can you please check ?

**Review by juherr — APPROVED** [review]

**Pull request merged**